Use custom typehint formatter

jl-wynen opened the pull request:
Fixes [#61](https://github.com/scipp/sciline/issues/61)

The docs don't show type aliases, so the refs that this generates don't work. I added them regardless for when we (hopefully) find a way to show aliases. I added the underlying type so that users can find out what types are actually required.

The result is still quite verbose. It could be made more compact by removing the module name. But then it is harder to find the definitions of the types.

With this change, this:
![before](https://github.com/scipp/copier_template/assets/11393224/17ceaff3-71c1-454d-aa9e-90281af58a22)

becomes this:
![after](https://github.com/scipp/copier_template/assets/11393224/26d7ef4b-3c72-4505-95cc-b91db1d42dc5)

---

But it does not handle `TypeVar`. So this:
![before2](https://github.com/scipp/copier_template/assets/11393224/809483c0-438a-46dc-b0d2-619163a4f06b)

becomes this:
![after2](https://github.com/scipp/copier_template/assets/11393224/b36765a1-73b8-4844-ad54-9d8dfc6f9540)

SimonHeybrock commented on template/docs/conf.py.jinja:
```diff
@@ -73,6 +73,34 @@ napoleon_type_aliases = {
+    if (supertype := getattr(annotation, '__supertype__', None)) is not None:
+        projectname = '{% if namespace_package %}{{namespace_package}}.{% endif %}{{ projectname.removeprefix(namespace_package) }}'
+
+        if not annotation.__module__.startswith(projectname):
+            return None
```
Why do nothing in this case? What if we depend on another package that uses aliases?

jl-wynen replied:
I was cautious because I don't know other uses of `NewType`. And I thought the sphinx people must have a reason for their solution. But I can't think of a case right now where this would break something.

SimonHeybrock replied:
Ok, let us leave it for now!

SimonHeybrock reviewed (commented):
Looks good, I don't think it is too verbose.

Question: Can this be adapted so it also works nicely with Python 3.12 type aliases (I have not tried how those are rendered currently)?

jl-wynen commented:
> Question: Can this be adapted so it also works nicely with Python 3.12 type aliases (I have not tried how those are rendered currently)?

It gets rendered in a bad way (my function does not apply to this case, so this is the default):
![py312](https://github.com/scipp/copier_template/assets/11393224/9ec696dd-ca52-4422-b436-55c9cf6b2647)

This uses
```python
type Filename = str;
"""The name of a file."""

Mode = NewType("Mode", str)
"""File open mode"""

Flags: TypeAlias = int
""""File flags"""
```

Interestingly, the link to `Mode` works now. Maybe something changes in python 3.12?

I'll try and improve the rendering.

jl-wynen commented:
I added handling of `TypeAliasType`. With the above dummy example, this produces the following. (I added a type parameter to `Filename` for testing. I there is no param, the `[T]` disappears.)
![p312-after](https://github.com/scipp/copier_template/assets/11393224/ca8eded0-0bd4-4725-8f46-8aa131c85d52)

SimonHeybrock commented on template/docs/conf.py.jinja:
```diff
@@ -134,7 +138,7 @@ html_theme_options = {
     "external_links": [
         {% for link in related_projects.replace(' ', '').split(',')|sort -%}
             {"name": "{{ link }}", "url": "https://scipp.github.io{% if link == 'Scipp' %}{% else %}/{{ link|lower }}{% endif %}"},
-        {% endfor %}],
+    {% endfor %}],
```
Intentional?

jl-wynen replied:
Yes, black changed the corresponding line in the rendered `conf.py`.

SimonHeybrock commented on template/docs/_typehints.py:
```diff
@@ -0,0 +1,82 @@
+    return f':{kind}:`{label} <{target}>`'
+
+
+def _link(ty: type, kind: str, type_params: Optional[tuple[type, ...]] = None) -> str:
+    if ty.__module__ == 'builtins':
+        target = ty.__name__
+    else:
+        target = f'{ty.__module__}.{ty.__name__}'
+    label = _typehint_aliases.get(target, target)
+    if type_params:
+        label += f'[{", ".join(ty.__name__ for ty in type_params)}]'
+    return f':{kind}:`{label} <{target}>`'
```
If I get this correctly we get working links now?

jl-wynen replied:
We do for uses of `NewType`. It seems that either with the latest Python or latest Sphinx, they are shown in the docs and then we also get links. It doesn't work with `type Filename = str` as the example shows, though.
